p4_symbolic/symbolic/util.cc

```diff
@@ -37,189 +43,149 @@ bool Z3BooltoBool(Z3_lbool z3_bool) {
-          model.eval(header.outer_ipv4_dst, true).get_numeral_int(),
-          model.eval(header.outer_ipv6_dst_upper, true).get_numeral_int(),
-          model.eval(header.outer_ipv6_dst_lower, true).get_numeral_int(),
-          model.eval(header.outer_protocol, true).get_numeral_int(),
-          model.eval(header.outer_dscp, true).get_numeral_int(),
-          model.eval(header.outer_ttl, true).get_numeral_int(),
-
-          model.eval(header.inner_ipv4_dst, true).get_numeral_int(),
-          model.eval(header.inner_ipv6_dst_upper, true).get_numeral_int(),
-          model.eval(header.inner_ipv6_dst_lower, true).get_numeral_int(),
-          model.eval(header.inner_protocol, true).get_numeral_int(),
-          model.eval(header.inner_dscp, true).get_numeral_int(),
-          model.eval(header.inner_ttl, true).get_numeral_int(),
-
-          model.eval(header.icmp_type, true).get_numeral_int(),
-          model.eval(header.vid, true).get_numeral_int()};
-}
-
-z3::expr MergeExpressionWithCondition(const z3::expr &original,
-                                      const z3::expr &changed,
-                                      const z3::expr &condition) {
-  if (z3::eq(original, changed)) {
-    return original;
-  }
-  return z3::ite(condition, changed, original);
-}
-
 }  // namespace
 
-SymbolicPerPacketState FreeSymbolicPacketState() {
-  // Port variables.
-  z3::expr ingress_port = Z3Context().int_const("ingress_port");
-
+gutil::StatusOr<SymbolicPerPacketState> FreeSymbolicPacketState(
+    const google::protobuf::Map<std::string, ir::HeaderType> &headers) {
   // Packet header variables.
-  SymbolicHeader header = {
-      Z3Context().int_const("ingress_eth_src"),
-      Z3Context().int_const("ingress_eth_dst"),
-      Z3Context().int_const("ingress_eth_type"),
-
-      Z3Context().int_const("ingress_outer_ipv4_src"),
-      Z3Context().int_const("ingress_outer_ipv4_dst"),
-      Z3Context().int_const("ingress_outer_ipv6_dst_upper"),
-      Z3Context().int_const("ingress_outer_ipv6_dst_lower"),
-      Z3Context().int_const("ingress_outer_protocol"),
-      Z3Context().int_const("ingress_outer_dscp"),
-      Z3Context().int_const("ingress_outer_ttl"),
-
-      Z3Context().int_const("ingress_inner_ipv4_dst"),
-      Z3Context().int_const("ingress_inner_ipv6_dst_upper"),
-      Z3Context().int_const("ingress_inner_ipv6_dst_lower"),
-      Z3Context().int_const("ingress_inner_protocol"),
-      Z3Context().int_const("ingress_inner_dscp"),
-      Z3Context().int_const("ingress_inner_ttl"),
-
-      Z3Context().int_const("ingress_icmp_type"),
-      Z3Context().int_const("ingress_vid"),
-  };
-
-  // Default metadata.
+  SymbolicHeader header = headers::FreeSymbolicHeader();
+
+  // Metadata fields.
   SymbolicMetadata metadata;
-  metadata.insert({"standard_metadata.ingress_port", ingress_port});
-  metadata.insert({"standard_metadata.egress_spec", Z3Context().int_val(-1)});
+  for (const auto &[header_name, header_type] : headers) {
```

Comment:
This is where we look up the header type definitions from the bmv2 json, and create free variables to every field defined there, with the same bitwidth.

Reply:
The pipeline down stream is going to be copying metadata around, and assigning new expressions to the fields in it as the p4 program does. Trigged the copy assignment overloaded operation.